### ISSUE TITLE
SCIM 2 listing users with pagination gives wrong totalResults

### ISSUE THREAD

**Issue by bala-striva** [wso2/product-is#7320]
WSO2 IS 5.9

When using pagination totalResults not returning the total number of resources but the number of returned resources.

Query all users give back 6 users.

```
➜ https --verify no -a admin:admin :9443/scim2/Users                                   
HTTP/1.1 200 
Cache-Control: private
Content-Length: 1880
Content-Type: application/scim+json
Date: Mon, 20 Jan 2020 14:24:39 GMT
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Server: WSO2 Carbon Server
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block

{
    "Resources": [
        {
            "emails": [
                "admin@wso2.com"
            ],
            "groups": [
                {
                    "display": "admin",
                    "value": "ec92c9f0-925f-43a1-98d6-d2f1746139cd"
                }
            ],
            "id": "ef9880af-a757-488e-9acc-7dfd7a51447d",
            "meta": {
                "created": "2020-01-20T14:22:33Z",
                "lastModified": "2020-01-20T14:22:33Z"
            },
            "name": {
                "familyName": "admin",
                "givenName": "admin"
            },
            "userName": "admin"
        },
        {
            "id": "a642c954-e65c-48f9-bc49-dcf54a7e2555",
            "meta": {
                "created": "2020-01-20T14:24:15.807Z",
                "lastModified": "2020-01-20T14:24:15.807Z",
                "location": "https://localhost:9443/scim2/Users/a642c954-e65c-48f9-bc49-dcf54a7e2555",
                "resourceType": "User"
            },
            "name": {
                "familyName": "bala1"
            },
            "userName": "bala1"
        },
        {
            "id": "67953177-33a7-4434-bac1-e34a612bd859",
            "meta": {
                "created": "2020-01-20T14:24:20.278Z",
                "lastModified": "2020-01-20T14:24:20.278Z",
                "location": "https://localhost:9443/scim2/Users/67953177-33a7-4434-bac1-e34a612bd859",
                "resourceType": "User"
            },
            "name": {
                "familyName": "bala2"
            },
            "userName": "bala2"
        },
        {
            "id": "f670a596-7901-4cf4-bc0d-cd115fa64527",
            "meta": {
                "created": "2020-01-20T14:24:23.788Z",
                "lastModified": "2020-01-20T14:24:23.788Z",
                "location": "https://localhost:9443/scim2/Users/f670a596-7901-4cf4-bc0d-cd115fa64527",
                "resourceType": "User"
            },
            "name": {
                "familyName": "bala3"
            },
            "userName": "bala3"
        },
        {
            "id": "ac57c8bf-7b69-4fd5-b1c5-bba7e04e3f5a",
            "meta": {
                "created": "2020-01-20T14:24:27.159Z",
                "lastModified": "2020-01-20T14:24:27.159Z",
                "location": "https://localhost:9443/scim2/Users/ac57c8bf-7b69-4fd5-b1c5-bba7e04e3f5a",
                "resourceType": "User"
            },
            "name": {
                "familyName": "bala4"
            },
            "userName": "bala4"
        },
        {
            "id": "640ee422-8360-440a-9c45-64d040d95fc3",
            "meta": {
                "created": "2020-01-20T14:24:30.639Z",
                "lastModified": "2020-01-20T14:24:30.639Z",
                "location": "https://localhost:9443/scim2/Users/640ee422-8360-440a-9c45-64d040d95fc3",
                "resourceType": "User"
            },
            "name": {
                "familyName": "bala5"
            },
            "userName": "bala5"
        }
    ],
    "itemsPerPage": 6,
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:ListResponse"
    ],
    "startIndex": 1,
    "totalResults": 6
}
```

Using **startIndex** and **count** reports wrong **totalResults** as 3

```
➜ https --verify no -a admin:admin :9443/scim2/Users startIndex==0 count==3
HTTP/1.1 200 
Cache-Control: private
Content-Length: 1010
Content-Type: application/scim+json
Date: Mon, 20 Jan 2020 14:25:15 GMT
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Server: WSO2 Carbon Server
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-XSS-Protection: 1; mode=block

{
    "Resources": [
        {
            "emails": [
                "admin@wso2.com"
            ],
            "groups": [
                {
                    "display": "admin",
                    "value": "ec92c9f0-925f-43a1-98d6-d2f1746139cd"
                }
            ],
            "id": "ef9880af-a757-488e-9acc-7dfd7a51447d",
            "meta": {
                "created": "2020-01-20T14:22:33Z",
                "lastModified": "2020-01-20T14:22:33Z"
            },
            "name": {
                "familyName": "admin",
                "givenName": "admin"
            },
            "userName": "admin"
        },
        {
            "id": "a642c954-e65c-48f9-bc49-dcf54a7e2555",
            "meta": {
                "created": "2020-01-20T14:24:15.807Z",
                "lastModified": "2020-01-20T14:24:15.807Z",
                "location": "https://localhost:9443/scim2/Users/a642c954-e65c-48f9-bc49-dcf54a7e2555",
                "resourceType": "User"
            },
            "name": {
                "familyName": "bala1"
            },
            "userName": "bala1"
        },
        {
            "id": "67953177-33a7-4434-bac1-e34a612bd859",
            "meta": {
                "created": "2020-01-20T14:24:20.278Z",
                "lastModified": "2020-01-20T14:24:20.278Z",
                "location": "https://localhost:9443/scim2/Users/67953177-33a7-4434-bac1-e34a612bd859",
                "resourceType": "User"
            },
            "name": {
                "familyName": "bala2"
            },
            "userName": "bala2"
        }
    ],
    "itemsPerPage": 3,
    "schemas": [
        "urn:ietf:params:scim:api:messages:2.0:ListResponse"
    ],
    "startIndex": 1,
    "totalResults": 3
}

```

**Comment by sarubi** [comment]
Hi @bala-striva , 
Thanks for reporting the behaviour.
We have addressed these concerns in our coming up releases such as to get the total count of users from DB. You can refer to the discussion mail thread [1] to keep the track on the progress.  
We would like to know what are the limitations that you are facing to implement your use-case which relies on the total count of users from DB. 

[1] http://wso2-oxygen-tank.10903.n7.nabble.com/Issue-in-USER-and-ROLES-count-implementation-td160975.html

**Comment by bala-striva** [comment]
Hi @sarubi ,

We would like to create a paginated list of Users and for that to know the number of pages we would need the **totalResults** to be accurate.

**Comment by IsuraD** [comment]
This is fixed for JDBCUserStores with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/243

**Comment by AnuradhaSK** [comment]
We decided not to implement the backend support to count the total number of users for LDAP user stores by loading all users into a data structure due to the performance issues. This will be clearly mentioned in the documentation with the reason of unsupported functionality in LDAP.  [DOC ISSUE: https://github.com/wso2/docs-is/issues/1495]
Since we already have a method to calculate the total result of users for JDBC user stores, we can obtain the correct total result in response to JDBC user stores.

We decided to handle the pagination from the frontend irrespective of the user store which is used. Since the pagination is an older mechanism and the new trend is infinitely scrolling, where we suggest doing in-memory pagination. 
For example: If we want to have 10 users per page, query, and get around 50 users from the database and buffer it at the frontend. Then the application can display up to 5 pages initially.
